lib/onelogin/ruby-saml/metadata.rb

```diff
@@ -117,23 +142,33 @@ def generate(settings, pretty_print=false, valid_until=nil, cache_duration=nil)
         #  <md:RoleDescriptor xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:query="urn:oasis:names:tc:SAML:metadata:ext:query" xsi:type="query:AttributeQueryDescriptorType" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol"/>
         #  <md:XACMLAuthzDecisionQueryDescriptor WantAssertionsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol"/>
 
-        meta_doc << REXML::XMLDecl.new("1.0", "UTF-8")
+        sp_sso
+      end
 
-        # embed signature
-        if settings.security[:metadata_signed] && settings.private_key && settings.certificate
-          private_key = settings.get_sp_key
-          meta_doc.sign_document(private_key, cert, settings.security[:signature_method], settings.security[:digest_method])
-        end
+      # can be overridden in subclass
+      def add_extras(root, _settings)
+        root
+      end
+
+      def embed_signature(meta_doc, settings)
+        return unless settings.security[:metadata_signed] && settings.private_key && settings.certificate
+
+        private_key = settings.get_sp_key
+        cert = settings.get_sp_cert
+        meta_doc.sign_document(private_key, cert, settings.security[:signature_method], settings.security[:digest_method])
```

Comment:
meta_doc.sign_document(settings.get_sp_key, settings.get_sp_cert, settings.security[:signature_method], settings.security[:digest_method])

Reply:
Done. Note the first line is `unless` not `if`, but otherwise exactly as you said.

Reply:
you are right ;)